src/main/java/ch/njol/skript/classes/data/JavaClasses.java

```diff
@@ -621,11 +239,483 @@ public void deserialize(String o, Fields f) {
 					public String deserialize(String s) {
 						return s;
 					}
-					
+
 					@Override
 					public boolean mustSyncDeserialization() {
 						return false;
 					}
 				}));
 	}
+
+	/**
+	 * Converts a string to a number formatted as an integer.
+	 * <p>
+	 * Applies {@code stringToNumber} for parsing the number, then tries to
+	 * convert radians to degrees if the string contains a radian group.
+	 * If the string could be parsed, applies {@code converter} to convert
+	 * the number to the desired type.
+	 * </p>
+	 *
+	 * @param string The string with the possible number.
+	 * @param stringToNumber The function to parse the number, e.g. {@link Integer#parseInt(String)}.
+	 * @param converter The function to convert the number to the desired type, e.g. {@link Number#intValue()}.
+	 * @return The parsed string, or null if the string could not be parsed.
+	 */
+	@Contract(pure = true)
+	private static <T extends Number> @Nullable T convertIntegerFormatted(
+		String string,
+		Function<String, Number> stringToNumber,
+		Function<Number, T> converter
+	) {
+		Matcher matcher = INTEGER_PATTERN.matcher(string);
+
+		if (!matcher.matches())
+			return null;
+
+		String number = matcher.group("num");
+		if (matcher.group("rad") != null) {
+			try {
+				return converter.apply(Math.toDegrees(stringToNumber.apply(number).doubleValue()));
+			} catch (NumberFormatException ignored) {
+			}
+		} else {
+			try {
+				return converter.apply(stringToNumber.apply(number));
+			} catch (NumberFormatException ignored) {
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Converts a string to a number formatted as an decimal.
+	 * <p>
+	 * Applies {@code stringToNumber} for parsing the number.
+	 * If the number is a percentage, it gets parsed as a double between 0-1.
+	 * Then tries to convert radians to degrees if the string contains a radian group.
+	 * If the string could be parsed, applies {@code converter} to convert the number to the desired type.
+	 * If the number is not valid by {@code isValid}, returns null.
+	 * </p>
+	 *
+	 * @param string The string with the possible number.
+	 * @param stringToNumber The function to parse the number, e.g. {@link Integer#parseInt(String)}.
+	 * @param fromPercentage A function that divides the value by 100.
+	 * @param converter The function to convert the number to the desired type, e.g. {@link Number#intValue()}.
+	 * @param isValid The function to check if the number is valid, e.g. {@link Double#isNaN()}.
```

Comment:
does this need to be done in the function or would it really harm anything to just return Number and check validity after?